### ISSUE TITLE
Try poetry

### ISSUE THREAD

**Issue by csernazs** [csernazs/pytest-httpserver#47]
Look at poetry, how much it developed during the years and how it feels to abandon setup.py and its friends.

**Comment by mhokhalifa** [comment]
We have been using poetry for a while now in our projects and it's made life much easier. If you decide for it I could gladly take care of the migration.

**Comment by csernazs** [comment]
Thanks for the help! I haven't used poetry for anything real, I played with it but that's all. First I need to convince myself that it is the right direction (in other words: please don't submit a PR for this yet), then I'll probably ask for your help to design it.

**Comment by cristianMeli** [comment]
Hi @csernazs!! First, thank you for this great plugin!
In my work we recently support pyproyect.toml + poetry, those guys help us to unify different dependency declaration files and to avoid conflicts between dependency versions (among others). We are working on adopting these in our applications.
Would you mind if I work on this issue?

**Comment by csernazs** [comment]
Hi,

Thanks :)

I looked at poetry several times and for pytest-httpserver it seems to be not giving so much. 
Currently there's a Makefile and a plain old simple setup.py file which does the work.  Pytest-httpserver has werkzeug as the only requirement, and by design I'm not pinning any dependencies. Werkzeug has a very stable API and I have no reason to fix any versions - it must be done at the application side which is using my library (and it should fix pytest-httpserver version along with Werkzeug).

As far as I remember, I had two issues with poetry:

- Testing with the latest dependencies: To test how the new werkzeug works with pytest-httpserver, currently I need to run the tests, as when installing the venv it will pick up the latest version. I found that poetry always create a lock file, which is not needed for me (and I think it causes more trouble).
- Multiple python versions: pytest-httpserver supports multiple python versions, but poetry requires tho specify the python version in pyproject.toml, so I cannot specify only one. I can specify the oldest python I still support, but in such case it won't pick up the latest Werkzeug. If I specify the latest, I cannot test with the old one. I run tests with gitlab CI for all the supported python versions, as I want to ensure that the code runs fine on all pythons.

Honestly, I like poetry and the whole idea behind it (and I encourage everyone using it at the application), but for me it seems like a pain for library authors like me. Last time I looked at how other libraries handle it and interestingly I haven't found any library which uses it - if you know anyone who has a good CI for their library, please let me know.

I'm open to migrate to it and appreciate any work you may put into this, but the above problems needs to be solved first.

**Comment by cristianMeli** [comment]
Good Monday! Thanks for all the info, I understand that the migration could not give you so much.

- Testing with the latest dependencies: yes, poetry creates a lock file. But if you run [poetry update](https://python-poetry.org/docs/cli/#update) it will update the lock file and the dependencies installed.

- Multiple Python versions: here I don't understand so much. You can set all python versions that you support in pyproject.toml. In your dependencies you could set, for example python = ">=3.6.1,<4.0".
I do not understand when you said  _"I can specify the oldest python I still support, but in such case it won't pick up the latest Werkzeug"_

**Comment by csernazs** [comment]
Ok, I think I'll give it another try - and if it does not work, I'll write down my findings. :)

If I remember correctly, when I tried poetry I committed the lock file, maybe that was not a good idea. It wired in the python version it saw on the current system, so I could not run the tests with other python version as it was not able to install the package dependencies due to python version mismatch.
As far as I remember it created a wheel which also included some wired-in python version but I'm not entirely sure about this.

Anyway, I will give poetry another try and get back with the results.

**Comment by csernazs** [comment]
Hi all,

I think I've made some progress in moving towards poetry.

Few remarks:
- I tend to ignore poetry.lock file in my repo, as I want to use the latest packages. This is how this project worked, I'm not entirely sure that this is a great idea, so will think about it. Having "*" specified everywhere is a big lie as pytest-httpserver won't work with the initial version of werkzeug, I'm totally sure.
- wheel now contains the tests, examples, etc. I wanted these to be included only in sdist but it's fine by me.
- sdist does not contain the "pytest_httpserver.egg-info" directory which contains the metadata (although PKG-INFO is there but in the package root). It seems in a recent pip this causes no problems (I could install the sdist also)
- there are minor diferences between the setup.py generated builds, I had to look at very thoroughly, but I still worring a bit about breaking everything. :)
- doc build and github CI needed to be adjusted (finally I could remove requirements-doc.txt)

As I think you are far more experienced with poetry, could you please have a look at my branch?
https://github.com/csernazs/pytest-httpserver/tree/wip-zsolt-poetry
e7f6c97cc23ac11d70bf0499df7cab80f14a2994 is my most recent commit

And also, any thoughts on the remarks above is highly welcome.

**Comment by cristianMeli** [comment]
Good Friday!!

Good, I think is great. 
I have a question:
* Why did you put the dev dependencies in an extra group and not in `[tool.poetry.dev-dependencies]`? Is it because is not possible to install only the dev dependencies?

Then I have some questions that are not related to pyproject, just for curiosity:
* Why do you include [here](https://github.com/csernazs/pytest-httpserver/blob/8ddfbf4d87a4ac36e49979a8e0376eb4950d8e10/pyproject.toml#L25) the tests in the package?

* Why did you not pin the dependencies versions? Previously, you said that is for design, and I'm curious about it. 
I think that pin the dependencies version, first you are explicit with which versions the project is compatible. Also, you ensure that the project will not break if there are breaking changes when a major version is released. 

Greetings!

**Comment by csernazs** [comment]
Hi,

> Why did you put the dev dependencies in an extra group and not in [tool.poetry.dev-dependencies]? Is it because is not possible to install only the dev dependencies?

I wanted to make a wheel (and sdist) which is as much closer to the current one (made by setup.py) as possible. The original one had this dev and test sections and I wanted to have the same. tool.poetry.dev-dependencies does not do that (although from the poetry perspective it worked for me fine).

> Why do you include here the tests in the package?
This came from comparing the sdist, as the tests directory was missing from the one which poetry produced and was included in the old one made without poetry. By this addition it is included in the wheel also, but when installing the package the tests directory is not installed to anywhere (same applies to every other files which are not package data, eg. README.md).
I would like to have the sdist as kind of a snapshot of the git revision so it can be used to start developing on the project, which means it should be sufficient to run the tests also.

> Why did you not pin the dependencies versions? 

There are two things here. First, specifying versions in pyproject.toml: I would like to have the code using pytest-httpserver to pin the dependencies as I'm saying that pytest-httpserver can work "pretty much" with all versions of "modern" werkzeug.  This does not mean that it works with ancient versions, so I hope everyone using the latest one. On the other hand, I should have provide the absolute mininal version here which can work with pytest-httpserver. I think I'll do that.

Second, fixing the versions in the lock file: I would like to test with the latest versions so I find having to update the lock file is just an unneeded thing. I have many dev dependencies, if those break, I need to follow it, but this won't cause any issues for the users (as they are not installing the dev dependencies). Also, having the "latest" version of werkzeug specified in the lock file depends on the python version I think. As I'm testing with different python versions, I think it is not possible to have a single lock file with a single release of werkzeug working on all python versions (or if there's such version of werkzeug exists, that won't be the latest one).

**Comment by cristianMeli** [comment]
Hi!! Thank you for the reply and all info!

**Comment by csernazs** [comment]
Alright, so I think it will be good to have poetry.lock file. It pretty much contains the depdendencies about python cores as well so it will be ok for testing multiple python versions, i hope. And this means that my software stack will be deterministic, so I need some bot (dependabot for example) to update it  (eg. to test with the latest werkzeug).

I also updated the version constraint for werzeug, so it will be in-sync with my python version constraints (I need >=3.6 so I picked the werkzeug 2.0.0 as the minimal as they dropped 3.5 and 2 support in it).
